README
This is a simple Java line application to output today's date. "Yahrzeit" is
"time of year" in Yiddish, used to remember important dates; I added support
for a feature to output events on a given day largely for the purposes of
the actual yahrzeit.

To use the remembrance feature, create a readable file called ~/.calendar.

~/.calendar has a series of lines in it, with the format:

   date,description

An example .calendar might be:

   Tammuz 20,Birthday for Sam Hill
   July 17,Anniversary
   Ramadan 4,Birthday, Ali

If the date matches the current day in the Hebrew, Gregorian, or Islamic
calendars, the description is output.

Note that the spelling of the Hebrew months is according to the Hebrew
Academy; thus, "Tammuz" is used (as opposed to "Tamuz"), although I used
Cheshvan instead of Marhesvan as the Academy recommends.

(I was using Marhesvan, but then realized that when I thought of it, it was
as "Cheshvan." A TODO is to make it so month names are customizable such
that you can use whichever spellings you prefer instead of relying on my
inconsistent and non-normalized transliterations.)

For Adar, the number is included; thus, "Adar 5" will not match, but "Adar I
5" will.

Note that for events in Adar II, matching will be inconsistent, because Adar
II is intercalatory; it's not present in every year.  The matching algorithm
is brute-force, not an actual matching algorithm based on calendar dates.

The algorithms are taken from
http://people.sc.fsu.edu/~jburkardt/cpp_src/calendar_rd/calendar_rd.cpp. 
The date support was originally taken from Joda-time, but this created a
large executable jar file, and using my own date classes was shorter (and
more clear).  (I also didn't want to write a Hebrew chronology for Joda-time
or Threeten, because I don't think I understand what the chronologies need
to support.) 

NEWS
-------

July 5, 2012 (Tamuz 15, 5772)

This should now build and run.

-------

src/main/java/com/enigmastation/yahrzeit/Yahrzeit.java
package com.enigmastation.yahrzeit;

import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Scanner;

public class Yahrzeit {
    @Parameter(names = {"-s", "--since"})
    private boolean since = false;

    @Parameter(names = {"-o", "--output"})
    private boolean outputToday = true;

    @Parameter(names = {"-h", "--hebrew"})
    private boolean showHebrew = true;

    @Parameter(names = {"-i", "--islamic"})
    private boolean showIslamic = false;

    @Parameter(names = {"-g", "--gregorian"})
    private boolean showGregorian = false;

    public void run() {
        Calendar c = Calendar.getInstance();
        GregorianDate gregorian = new GregorianDate(c);
        HebrewDate hebrew = new HebrewDate(c);
        IslamicDate islamic = new IslamicDate(c);

        List<String> dateMatches = new ArrayList<>();

        dateMatches.add(toString(hebrew));
        dateMatches.add(toString(islamic));
        dateMatches.add(toString(gregorian));

        if (outputToday) {
            show(showHebrew, hebrew);
            show(showIslamic, islamic);
            show(showGregorian, gregorian);
        }
        File calendarFile = new File(System.getProperty("user.home") +
                System.getProperty("file.separator") + ".calendar");
        try {
            FileReader fr = new FileReader(calendarFile);
            BufferedReader br = new BufferedReader(fr);
            Scanner scanner = new Scanner(br);
            String input;
            while (scanner.hasNextLine()) {
                input = scanner.nextLine();
                for (String date : dateMatches) {
                    if (input.trim().startsWith(date)) {
                        System.out.println(input.substring(date.length() + 1).trim());
                    }
                }
            }
        } catch (FileNotFoundException ignored) {
            // this is okay, we just don't output anything if the file doesn't exist
        }

    }

    private String toString(YahrzeitDate date) {
        return date.getMonthName() + " " + date.getDay();
    }

    private void show(boolean showDate, YahrzeitDate date) {
        if (showDate) {
            System.out.println(date.getMonthName() + " " + date.getDay() + ", " + date.getYear());
        }
    }

    public static void main(String[] args) {
        Yahrzeit y = new Yahrzeit();
        new JCommander(y, args);
        y.run();
    }
}


src/main/java/com/enigmastation/yahrzeit/HebrewDate.java
package com.enigmastation.yahrzeit;

import java.util.Calendar;

public class HebrewDate extends YahrzeitDate {

    final static int epoch = -1373429;    // Absolute date of start of Hebrew calendar

    public HebrewDate(int month, int day, int year) {
        this.month = month;
        this.day = day;
        this.year = year;
    }

    public HebrewDate(Calendar c) {
        initFromAbsoluteDate(toAbsolute(c));
    }

    public HebrewDate() {
        Calendar c = Calendar.getInstance();
        int absoluteDay = toAbsolute(c);
        initFromAbsoluteDate(absoluteDay);
    }

    @Override
    public String getMonthName() {
        String monthNames[] = {"Nisan", "Iyyar", "Sivan", "Tammuz", "Av",
                "Elul", "Tishri", "Cheshvan", "Kislev", "Tevet", "Shvat",
                "Adar I", "Adar II"
        };
        return monthNames[month - 1];
    }

    @Override
    public int toAbsolute() {
        int dayInYear = getDay();    // Days so far this month.
        if (getMonth() < 7) {   // Before Tishri, so add days in prior months
            // this year before and after Nisan.
            int m = 7;
            while (m <= (lastMonth(getYear()))) {
                dayInYear = dayInYear + lastDayOfMonth(m, getYear());
                m++;
            }
            m = 1;
            while (m < getMonth()) {
                dayInYear = dayInYear + lastDayOfMonth(m, getYear());
                m++;
            }
        } else {   // Add days in prior months this year
            int m = 7;
            while (m < getMonth()) {
                dayInYear = dayInYear + lastDayOfMonth(m, getYear());
                m++;
            }
        }
        return (dayInYear + (elapsedDays(getYear())    // Days in prior years.
                + epoch));    // Days elapsed before absolute date 1.
    }

    private void initFromAbsoluteDate(int d) {
        year = (d + epoch) / 366;
        while (d >= new HebrewDate(7, 1, year + 1).toAbsolute())
            year++;
        // Search forward for month from either Tishri or Nisan.
        if (d < new HebrewDate(1, 1, year).toAbsolute())
            month = 7;        //  Start at Tishri
        else
            month = 1;        //  Start at Nisan
        while (d > new HebrewDate(getMonth(), (lastDayOfMonth(getMonth(), year)), year).toAbsolute())
            month++;
        // Calculate the day by subtraction.
        day = d - new HebrewDate(getMonth(), 1, year).toAbsolute() + 1;
    }

    static boolean isLeapYear(int year) {
        return (((7 * year) + 1) % 19) < 7;
    }

    static int lastMonth(int year) {
        return (isLeapYear(year)) ? 13 : 12;
    }

    static int elapsedDays(int year) {
        // Number of days elapsed from the Sunday prior to the start of the
// Hebrew calendar to the mean conjunction of Tishri of Hebrew year.

        int monthsElapsed = (235 * ((year - 1) / 19))    // Months in complete cycles so far.
                + (12 * ((year - 1) % 19))    // Regular months in this cycle.
                + (7 * ((year - 1) % 19) + 1) / 19;    // Leap months this cycle
        int partsElapsed = 204 + 793 * (monthsElapsed % 1080);
        int h = 5 + 12 * monthsElapsed + 793 * (monthsElapsed / 1080)
                + partsElapsed / 1080;
        int conjunctionDay = 1 + 29 * monthsElapsed + h / 24;
        int conjunctionParts = 1080 * (h % 24) + partsElapsed % 1080;
        int alternativeDay;
        if ((conjunctionParts >= 19440)    // If new moon is at or after midday,
                || (((conjunctionDay % 7) == 2)    // ...or is on a Tuesday...
                && (conjunctionParts >= 9924)    // at 9 hours, 204 parts or later...
                && !(isLeapYear(year)))    // ...of a common year,
                || (((conjunctionDay % 7) == 1)    // ...or is on a Monday at...
                && (conjunctionParts >= 16789)    // 15 hours, 589 parts or later...
                && (isLeapYear(year - 1))))    // at the end of a leap year
            // Then postpone Rosh HaShanah one day
            alternativeDay = conjunctionDay + 1;
        else
            alternativeDay = conjunctionDay;
        if (((alternativeDay % 7) == 0)    // If Rosh HaShanah would occur on Sunday,
                || ((alternativeDay % 7) == 3)    // or Wednesday,
                || ((alternativeDay % 7) == 5))    // or Friday
            // Then postpone it one (more) day
            return (1 + alternativeDay);
        else
            return alternativeDay;
    }

    static int daysInYear(int year) {
        return ((elapsedDays(year + 1))
                - (elapsedDays(year)));
    }

    static boolean longHeshvan(int year) {
        return ((daysInYear(year) % 10) == 5);
    }

    static boolean shortKislev(int year) {
        return ((daysInYear(year) % 10) == 3);
    }

    static int
    lastDayOfMonth(int month, int year) {
// Last day of month in Hebrew year.

        if ((month == 2) || (month == 4) || (month == 6)
                || ((month == 8) && !(longHeshvan(year)))
                || ((month == 9) && shortKislev(year)) || (month == 10)
                || ((month == 12) && !(isLeapYear(year))) || (month == 13))
            return 29;
        else
            return 30;
    }

}



src/main/java/com/enigmastation/yahrzeit/IslamicDate.java
package com.enigmastation.yahrzeit;

import java.util.Calendar;

public class IslamicDate extends YahrzeitDate {
    private static final int epoch = 227014;

    static boolean isLeapYear(int year) {
        // True if year is an Islamic leap year

        return (((11 * year) + 14) % 30) < 11;
    }

    static int
    lastDayOfIslamicMonth(int month, int year) {
        // Last day in month during year on the Islamic calendar.
        if (((month % 2) == 1) || ((month == 12) && isLeapYear(year)))
            return 30;
        else
            return 29;
    }

    public IslamicDate(int month, int day, int year) {
        this.month = month;
        this.day = day;
        this.year = year;
    }

    public IslamicDate(Calendar c) {
        initFromAbsolute(toAbsolute(c));
    }

    public IslamicDate(int d) {
        initFromAbsolute(d);
    }

    private void initFromAbsolute(int d) {
        if (d <= epoch) {   // Date is pre-Islamic
            month = 0;
            day = 0;
            year = 0;
        } else {
            // Search forward year by year from approximate year
            year = (d - epoch) / 355;
            while (d >= new IslamicDate(1, 1, year + 1).toAbsolute())
                year++;
            // Search forward month by month from Muharram
            month = 1;
            while (d > new IslamicDate(month, lastDayOfIslamicMonth(month, year),
                    year).toAbsolute())
                month++;
            day = d - new IslamicDate(month, 1, year).toAbsolute() + 1;
        }
    }

    @Override
    public String getMonthName() {
        String[] monthNames = {"Muharram", "Safar", "Rabi I", "Rabi II",
                "Jumada I", "Jumada II", "Rajab", "Sha'ban", "Ramadan", "Shawwal", "Dhu'l Qa'dah", "Dhu'l-Hijja"};
        return monthNames[getMonth()-1];
    }

    @Override
    public int toAbsolute() {
        return (getDay()            // days so far this month
                + 29 * (getMonth() - 1)    // days so far...
                + getMonth() / 2        //            ...this year
                + 354 * (getYear() - 1)    // non-leap days in prior years
                + (3 + (11 * getYear())) / 30    // leap days in prior years
                + epoch);    // days before start of calendar
    }
}


src/main/java/com/enigmastation/yahrzeit/YahrzeitDate.java
package com.enigmastation.yahrzeit;

import java.util.Calendar;

abstract public class YahrzeitDate {
    int year;
    int month;
    int day;

    abstract public String getMonthName();

    public int getYear() {
        return year;
    }

    public int getMonth() {
        return month;
    }

    public int getDay() {
        return day;
    }

    public abstract int toAbsolute();

    int toAbsolute(Calendar c) {
        int gm = c.get(Calendar.MONTH) + 1;
        int gd = c.get(Calendar.DAY_OF_MONTH);
        int gy = c.get(Calendar.YEAR);
        int N = gd;        // days this month
        for (int m = gm - 1; m > 0; m--)    // days in prior months this year
            N = N + lastDayOfGregorianMonth(m, getYear());
        return (N            // days this year
                + 365 * (gy - 1)    // days in previous years ignoring leap days
                + (gy - 1) / 4    // Julian leap days before this year...
                - (gy - 1) / 100    // ...minus prior century years...
                + (gy - 1) / 400);
    }

    int lastDayOfGregorianMonth(int month, int year) {
        // Compute the last date of the month for the Gregorian calendar.

        switch (month) {
            case 2:
                if ((((year % 4) == 0) && ((year % 100) != 0)) || ((year % 400) == 0))
                    return 29;
                else
                    return 28;
            case 4:
            case 6:
            case 9:
            case 11:
                return 30;
            default:
                return 31;
        }
    }

}


src/main/java/com/enigmastation/yahrzeit/GregorianDate.java
package com.enigmastation.yahrzeit;

import java.util.Calendar;

public class GregorianDate extends YahrzeitDate {
    public GregorianDate(Calendar c) {
        month=c.get(Calendar.MONTH);
        day=c.get(Calendar.DAY_OF_MONTH);
        year=c.get(Calendar.YEAR);
    }

    @Override
    public String getMonthName() {
        String monthNames[] = {"January", "February", "March", "April", "May", "June",
                "July", "August", "September", "October", "November", "December"};
        return monthNames[getMonth()];
    }

    @Override
    public int toAbsolute() {
        int N = day;        // days this month
        for (int m = month - 1; m > 0; m--)    // days in prior months this year
            N = N + lastDayOfGregorianMonth(m, year);
        return (N            // days this year
                + 365 * (year - 1)    // days in previous years ignoring leap days
                + (year - 1) / 4    // Julian leap days before this year...
                - (year - 1) / 100    // ...minus prior century years...
                + (year - 1) / 400);    // ...plus prior years divisible by 400
    }
}


src/yahrzeit.cpp
// The following C++ code is translated from the Lisp code
// in ``Calendrical Calculations'' by Nachum Dershowitz and
// Edward M. Reingold, Software---Practice & Experience,
// vol. 20, no. 9 (September, 1990), pp. 899--928.
// This code is in the public domain, but any use of it
// should publically acknowledge its source.
// Classes GregorianDate, JulianDate, IsoDate, IslamicDate,
// and HebrewDate
#include <fstream>
#include <sstream>
#include <sys/types.h>
#include <pwd.h>
#include <string>
#include <math.h>
#include <ctime>
#include <iostream>
#include <unistd.h>
#include <stdlib.h>

using namespace std;

class IsoDate;

const char *DayName[7] =
{   "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday",
    "Friday", "Saturday"
};

string MonthNames[] = { "January", "February", "March", "April", "May",
                        "June", "July", "August", "September", "October", "November", "December"
                      };

// Absolute dates

// "Absolute date" means the number of days elapsed since the Gregorian date
// Sunday, December 31, 1 BC. (Since there was no year 0, the year following
// 1 BC is 1 AD.) Thus the Gregorian date January 1, 1 AD is absolute date
// number 1.

int
XdayOnOrBefore (int d, int x)
{
// Absolute date of the x-day on or before absolute date d.
// x=0 means Sunday, x=1 means Monday, and so on.

    return (d - ((d - x) % 7));
}

//  Gregorian dates

int
LastDayOfGregorianMonth (int month, int year)
{
// Compute the last date of the month for the Gregorian calendar.

    switch (month)
    {
    case 2:
        if ((((year % 4) == 0) && ((year % 100) != 0)) || ((year % 400) == 0))
            return 29;
        else
            return 28;
    case 4:
    case 6:
    case 9:
    case 11:
        return 30;
    default:
        return 31;
    }
}

class GregorianDate
{
private:
    int year;			// 1...
    int month;			// 1 == January, ..., 12 == December
    int day;			// 1..LastDayOfGregorianMonth(month, year)

public:
    GregorianDate ()
    {
        time_t t = time (0);	// get time now
        struct tm *now = localtime (&t);
        month = now->tm_mon + 1;
        day = now->tm_mday;
        year = now->tm_year + 1900;
    }

    GregorianDate (int m, int d, int y)
    {
        month = m;
        day = d;
        year = y;
    }

    GregorianDate (int d)
    {   // Computes the Gregorian date from the absolute date.

        // Search forward year by year from approximate year
        year = d / 366;
        while (d >= GregorianDate (1, 1, year + 1))
            year++;
        // Search forward month by month from January
        month = 1;
        while (d
                > GregorianDate (month, LastDayOfGregorianMonth (month, year),
                                 year))
            month++;
        day = d - GregorianDate (month, 1, year) + 1;
    }

    operator   int ()
    {   // Computes the absolute date from the Gregorian date.
        int N = day;		// days this month
        for (int m = month - 1; m > 0; m--)	// days in prior months this year
            N = N + LastDayOfGregorianMonth (m, year);
        return (N			// days this year
                + 365 * (year - 1)	// days in previous years ignoring leap days
                + (year - 1) / 4	// Julian leap days before this year...
                - (year - 1) / 100	// ...minus prior century years...
                + (year - 1) / 400);	// ...plus prior years divisible by 400
    }

    int GetMonth ()
    {
        return month;
    }
    string GetMonthName ()
    {
        return MonthNames[GetMonth () - 1];
    }
    int GetDay ()
    {
        return day;
    }
    int GetYear ()
    {
        return year;
    }

};

ostream & operator<< (ostream & c, GregorianDate d)
{
    c << d.GetMonth () << "/" << d.GetDay () << "/" << d.GetYear ();
    return c;
}

;

GregorianDate
NthXday (int n, int x, int month, int year, int day = 0)
// The Gregorian date of nth x-day in month, year before/after optional day.
// x = 0 means Sunday, x = 1 means Monday, and so on.  If n<0, return the nth
// x-day before month day, year (inclusive).  If n>0, return the nth x-day
// after month day, year (inclusive).  If day is omitted or 0, it defaults
// to 1 if n>0, and month's last day otherwise.
{
    if (n > 0)
    {
        if (day == 0)
            day = 1;		// default for positive n
        return GregorianDate ((7 * (n - 1))
                              + XdayOnOrBefore (6 +
                                                GregorianDate (month, day,
                                                        year), x));
    }
    else
    {
        if (day == 0)
            day = LastDayOfGregorianMonth (month, year);
        ;				// default for negative n
        return GregorianDate ((7 * (n + 1))
                              +
                              XdayOnOrBefore (GregorianDate (month, day, year),
                                              x));
    }
}

// Julian dates

const int JulianEpoch = -2;	// Absolute date of start of Julian calendar

int
LastDayOfJulianMonth (int month, int year)
{
// Compute the last date of the month for the Julian calendar.
    switch (month)
    {
    case 2:
        if ((year % 4) == 0)
            return 29;
        else
            return 28;
    case 4:
    case 6:
    case 9:
    case 11:
        return 30;
    default:
        return 31;
    }
}

class JulianDate
{
private:
    int year;			// 1...
    int month;			// 1 == January, ..., 12 == December
    int day;			// 1..LastDayOfJulianMonth(month, year)

public:
    JulianDate (int m, int d, int y)
    {
        month = m;
        day = d;
        year = y;
    }

    JulianDate (int d)
    {   // Computes the Julian date from the absolute date.
        // Search forward year by year from approximate year
        year = (d + JulianEpoch) / 366;
        while (d >= JulianDate (1, 1, year + 1))
            year++;
        // Search forward month by month from January
        month = 1;
        while (d > JulianDate (month, LastDayOfJulianMonth (month, year), year))
            month++;
        day = d - JulianDate (month, 1, year) + 1;
    }

    operator   int ()
    {   // Computes the absolute date from the Julian date.

        int N = day;		// days this month
        for (int m = month - 1; m > 0; m--)	// days in prior months this year
            N = N + LastDayOfJulianMonth (m, year);
        return (N			// days this year
                + 365 * (year - 1)	// days in previous years ignoring leap days
                + (year - 1) / 4	// leap days before this year...
                + JulianEpoch);	// days elapsed before absolute date 1
    }

    int GetMonth ()
    {
        return month;
    }
    int GetDay ()
    {
        return day;
    }
    int GetYear ()
    {
        return year;
    }
    int absolute ();

};

ostream & operator<< (ostream & c, JulianDate d)
{
    c << d.GetMonth () << "/" << d.GetDay () << "/" << d.GetYear ();
    return c;
}

;

// ISO dates

class IsoDate
{
private:
    int year;			// 1...
    int week;			// 1..52 or 53
    int day;			// 1..7

public:
    IsoDate (int w, int d, int y)
    {
        week = w;
        day = d;
        year = y;
    }

    IsoDate (int d)
    {   // Computes the ISO date from the absolute date.
        year = GregorianDate (d - 3).GetYear ();
        if (d >= IsoDate (1, 1, year + 1))
            year++;
        if ((d % 7) == 0)
            day = 7;			// Sunday
        else
            day = d % 7;		// Monday..Saturday
        week = 1 + (d - IsoDate (1, 1, year)) / 7;
    }

    operator   int ()
    {   // Computes the absolute date from the ISO date.
        return XdayOnOrBefore (GregorianDate (1, 4, year), 1)	// days in prior years
               + 7 * (week - 1)		// days in prior weeks this year
               + (day - 1);		// prior days this week
    }

    int GetWeek ()
    {
        return week;
    }
    int GetDay ()
    {
        return day;
    }
    int GetYear ()
    {
        return year;
    }

};

ostream & operator<< (ostream & c, IsoDate d)
{
    c << d.GetWeek () << "/" << d.GetDay () << "/" << d.GetYear ();
    return c;
}

// Islamic dates

const int IslamicEpoch = 227014;	// Absolute date of start of Islamic calendar

int
IslamicLeapYear (int year)
{
// True if year is an Islamic leap year

    if ((((11 * year) + 14) % 30) < 11)
        return 1;
    else
        return 0;
}

int
LastDayOfIslamicMonth (int month, int year)
{
// Last day in month during year on the Islamic calendar.

    if (((month % 2) == 1) || ((month == 12) && IslamicLeapYear (year)))
        return 30;
    else
        return 29;
}

class IslamicDate
{
private:
    int year;			// 1...
    int month;			// 1..13 (12 in a common year)
    int day;			// 1..LastDayOfIslamicMonth(month,year)

public:
    IslamicDate (int m, int d, int y)
    {
        month = m;
        day = d;
        year = y;
    }

    IslamicDate (int d)
    {   // Computes the Islamic date from the absolute date.
        if (d <= IslamicEpoch)
        {   // Date is pre-Islamic
            month = 0;
            day = 0;
            year = 0;
        }
        else
        {
            // Search forward year by year from approximate year
            year = (d - IslamicEpoch) / 355;
            while (d >= IslamicDate (1, 1, year + 1))
                year++;
            // Search forward month by month from Muharram
            month = 1;
            while (d
                    > IslamicDate (month, LastDayOfIslamicMonth (month, year),
                                   year))
                month++;
            day = d - IslamicDate (month, 1, year) + 1;
        }
    }

    operator   int ()
    {   // Computes the absolute date from the Islamic date.
        return (day			// days so far this month
                + 29 * (month - 1)	// days so far...
                + month / 2		//            ...this year
                + 354 * (year - 1)	// non-leap days in prior years
                + (3 + (11 * year)) / 30	// leap days in prior years
                + IslamicEpoch);	// days before start of calendar
    }

    int GetMonth ()
    {
        return month;
    }
    int GetDay ()
    {
        return day;
    }
    int GetYear ()
    {
        return year;
    }

};

ostream & operator<< (ostream & c, IslamicDate d)
{
    c << d.GetMonth () << "/" << d.GetDay () << "/" << d.GetYear ();
    return c;
}

// Hebrew dates

const int HebrewEpoch = -1373429;	// Absolute date of start of Hebrew calendar

int
HebrewLeapYear (int year)
{
// True if year is an Hebrew leap year

    if ((((7 * year) + 1) % 19) < 7)
        return 1;
    else
        return 0;
}

int
LastMonthOfHebrewYear (int year)
{
// Last month of Hebrew year.

    if (HebrewLeapYear (year))
        return 13;
    else
        return 12;
}

int
HebrewCalendarElapsedDays (int year)
{
// Number of days elapsed from the Sunday prior to the start of the
// Hebrew calendar to the mean conjunction of Tishri of Hebrew year.

    int MonthsElapsed = (235 * ((year - 1) / 19))	// Months in complete cycles so far.
                        + (12 * ((year - 1) % 19))	// Regular months in this cycle.
                        + (7 * ((year - 1) % 19) + 1) / 19;	// Leap months this cycle
    int PartsElapsed = 204 + 793 * (MonthsElapsed % 1080);
    int HoursElapsed = 5 + 12 * MonthsElapsed + 793 * (MonthsElapsed / 1080)
                       + PartsElapsed / 1080;
    int ConjunctionDay = 1 + 29 * MonthsElapsed + HoursElapsed / 24;
    int ConjunctionParts = 1080 * (HoursElapsed % 24) + PartsElapsed % 1080;
    int AlternativeDay;
    if ((ConjunctionParts >= 19440)	// If new moon is at or after midday,
            || (((ConjunctionDay % 7) == 2)	// ...or is on a Tuesday...
                && (ConjunctionParts >= 9924)	// at 9 hours, 204 parts or later...
                && !(HebrewLeapYear (year)))	// ...of a common year,
            || (((ConjunctionDay % 7) == 1)	// ...or is on a Monday at...
                && (ConjunctionParts >= 16789)	// 15 hours, 589 parts or later...
                && (HebrewLeapYear (year - 1))))	// at the end of a leap year
        // Then postpone Rosh HaShanah one day
        AlternativeDay = ConjunctionDay + 1;
    else
        AlternativeDay = ConjunctionDay;
    if (((AlternativeDay % 7) == 0)	// If Rosh HaShanah would occur on Sunday,
            || ((AlternativeDay % 7) == 3)	// or Wednesday,
            || ((AlternativeDay % 7) == 5))	// or Friday
        // Then postpone it one (more) day
        return (1 + AlternativeDay);
    else
        return AlternativeDay;
}

int
DaysInHebrewYear (int year)
{
// Number of days in Hebrew year.

    return ((HebrewCalendarElapsedDays (year + 1))
            - (HebrewCalendarElapsedDays (year)));
}

int
LongHeshvan (int year)
{
// True if Heshvan is long in Hebrew year.

    if ((DaysInHebrewYear (year) % 10) == 5)
        return 1;
    else
        return 0;
}

int
ShortKislev (int year)
{
// True if Kislev is short in Hebrew year.

    if ((DaysInHebrewYear (year) % 10) == 3)
        return 1;
    else
        return 0;
}

int
LastDayOfHebrewMonth (int month, int year)
{
// Last day of month in Hebrew year.

    if ((month == 2) || (month == 4) || (month == 6)
            || ((month == 8) && !(LongHeshvan (year)))
            || ((month == 9) && ShortKislev (year)) || (month == 10)
            || ((month == 12) && !(HebrewLeapYear (year))) || (month == 13))
        return 29;
    else
        return 30;
}

class HebrewDate
{
private:
    int year;			// 1...
    int month;			// 1..LastMonthOfHebrewYear(year)
    int day;			// 1..LastDayOfHebrewMonth(month, year)

public:
    HebrewDate (int m, int d, int y)
    {
        month = m;
        day = d;
        year = y;
    }

    HebrewDate (int d)
    {   // Computes the Hebrew date from the absolute date.
        year = (d + HebrewEpoch) / 366;	// Approximation from below.
        // Search forward for year from the approximation.
        while (d >= HebrewDate (7, 1, year + 1))
            year++;
        // Search forward for month from either Tishri or Nisan.
        if (d < HebrewDate (1, 1, year))
            month = 7;		//  Start at Tishri
        else
            month = 1;		//  Start at Nisan
        while (d > HebrewDate (month, (LastDayOfHebrewMonth (month, year)), year))
            month++;
        // Calculate the day by subtraction.
        day = d - HebrewDate (month, 1, year) + 1;
    }

    operator   int ()
    {   // Computes the absolute date of Hebrew date.
        int DayInYear = day;	// Days so far this month.
        if (month < 7)
        {   // Before Tishri, so add days in prior months
            // this year before and after Nisan.
            int m = 7;
            while (m <= (LastMonthOfHebrewYear (year)))
            {
                DayInYear = DayInYear + LastDayOfHebrewMonth (m, year);
                m++;
            };
            m = 1;
            while (m < month)
            {
                DayInYear = DayInYear + LastDayOfHebrewMonth (m, year);
                m++;
            }
        }
        else
        {   // Add days in prior months this year
            int m = 7;
            while (m < month)
            {
                DayInYear = DayInYear + LastDayOfHebrewMonth (m, year);
                m++;
            }
        }
        return (DayInYear + (HebrewCalendarElapsedDays (year)	// Days in prior years.
                             + HebrewEpoch));	// Days elapsed before absolute date 1.
    }
    string GetMonthName ()
    {
        string monthNames[13] = { "Nisan", "Iyyar", "Sivan", "Tammuz", "Av",
                                  "Elul", "Tishri", "Cheshvan", "Kislev", "Tevet", "Shvat",
                                  "Adar I", "Adar II"
                                };
        return monthNames[GetMonth () - 1];
    }
    int GetMonth ()
    {
        return month;
    }
    int GetDay ()
    {
        return day;
    }
    int GetYear ()
    {
        return year;
    }

};

ostream & operator<< (ostream & c, HebrewDate d)
{
    c << d.GetMonth () << "/" << d.GetDay () << "/" << d.GetYear ();
    return c;
}

void checkMatch(string line, string exp) {
    if(line.size()>exp.size()) {
        if(line.substr(0,exp.size())==exp) {
            cout << line.substr(exp.size()+1) << endl;
        }
    }
}

int
main (int argc, char **argv)
{

    int m, d, y;
    GregorianDate now;
    int ea = now;
    HebrewDate hnow (ea);

    // get the user's home directory - we'll need it to check for calendar dates
    struct passwd *pw = getpwuid (getuid ());
    const char *homedir = pw->pw_dir;

    // now let's build two strings with today's day of month, one in
    // gregorian and the other in Hebrew.
    stringstream gregday, hebday, filename;

    gregday << now.GetMonthName () << " " << now.GetDay ();
    hebday << hnow.GetMonthName () << " " << hnow.GetDay ();

    cout << hebday.str () << ", " << hnow.GetYear () << " (" << gregday.
         str () << ")" << endl;

    // now we've output the date, let's look for calendar events!
    filename << pw->pw_dir << "/" << ".calendar";
    ifstream calendarFile;

    calendarFile.open (filename.str ().c_str ());
    if (calendarFile.is_open ())
    {
        string line;
        while (calendarFile.good ())
        {
            /*
            * now for every line, parse up to the first comma, and match against our
            *  two dates
            *
            * TODO Add support for other calendars too?
            */
            getline(calendarFile, line);
            checkMatch(line, hebday.str());
            checkMatch(line, gregday.str());
        }
    }

    return 0;
}
